TcUnit-Runner/VisualStudioInstance.cs

```diff
@@ -127,35 +131,61 @@ private void LoadDevelopmentToolsEnvironment(string visualStudioVersion, bool is
+            // Check if Version is installed
+            foreach (var possibleVersion in remoteManager.Versions)
             {
-                log.Info("Project has a pinned version: " + this.tcVersion);
-                foreach (var possibleVersion in remoteManager.Versions)
+                latestTcVersion = Char.ToString(possibleVersion[0]);
+                if (possibleVersion == this.tcVersion)
                 {
-                    if (possibleVersion == this.tcVersion)
-                    {
-                        isVersionAvailable = true;
-                        break;
-                    }
+                    isVersionAvailable = true;
                 }
-                if (isVersionAvailable)
-                    log.Info("The pinned TwinCAT version is available ");
-                else
+                // Check if optional Version is installed
+                if (possibleVersion == this.forceToThisTwinCatVersion && this.forceToThisTwinCatVersion != null)
                 {
-                    log.Error("The pinned TwinCAT version is not available ");
-                    throw new Exception("The pinned TwinCAT version is not available");
+                    isForceVersionAvailable = true;
+                    log.Info("The optional TC Version is available ");
                 }
             }
 
-            if (!isTcVersionPinned || isVersionAvailable)
+            // If the Version is Pinned, check if the Pinned Version is available
+            if (isTcVersionPinned && (String.IsNullOrEmpty(this.forceToThisTwinCatVersion)))
+            {
+
+                  if (isVersionAvailable)
+                  {
+                      log.Info("The pinned TC Version is available ");
+                  }
+                  else
+                  {
+                      log.Error("The pinned TC Version is not available ");
+                      throw new Exception("The pinned TC Version is not available");
+                  }
+             }
+
+
+            if (isTcVersionPinned && isVersionAvailable && (String.IsNullOrEmpty(this.forceToThisTwinCatVersion)))
             {
                 remoteManager.Version = this.tcVersion;
-                log.Info("Load TwinCAT version: "+ remoteManager.Version);
             }
+            else if (isForceVersionAvailable)
+            {
+                log.Info("TC Version is forced to " + this.forceToThisTwinCatVersion);
+                remoteManager.Version = this.forceToThisTwinCatVersion;
+            }
+            else if (!isForceVersionAvailable && (!String.IsNullOrEmpty(this.forceToThisTwinCatVersion)))
+            {
+                log.Error("The Optional TC Version "+ this.forceToThisTwinCatVersion + " is not available ");
+                throw new Exception("The Optional TC Version is not available");
+            }
+            else if (latestTcVersion != "")
+            {
+                remoteManager.Version = latestTcVersion;
+            }
```

Comment:
what happens if none of these cases are true?

Reply:
Good question: There is no attempt to start the Remote Manger

`if (latestTcVersion != "")`
If a TwinCat version is installed, this value is never NULL
It is more a security check that the variable has not been overwritten somewhere in the program.

I see 2 possibilities.
1. we remove the `if (latestTcVersion != ""),` if it is empty or invalid the Remote Manager will generate an exception

2. we write an else branch which generates an exception

Reply:
I like your option #2!

Reply:
maybe

```
 else
            {
               log.Error("Error: It´s not possible to run TwinCAT in this configuration (TC-Option / pinned TwinCAT Version / installed TwinCAT Version ");
               throw new Exception("Wrong configuration for this TwinCAT version ");
            }
```